File: nanobot/nanobot/interagent/tracing.py
"""
OpenTelemetry 分布式追踪模块
实现跨服务的分布式追踪能力

@module interagent.tracing
@version 1.0.0
"""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional, Callable, TypeVar, ParamSpec
from contextlib import asynccontextmanager, contextmanager
from functools import wraps
import os

# OpenTelemetry 类型（延迟导入以支持可选依赖）
try:
    from opentelemetry import trace
    from opentelemetry.trace import Span, StatusCode, SpanKind
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
    from opentelemetry.sdk.resources import Resource

    OPENTELEMETRY_AVAILABLE = True
except ImportError:
    OPENTELEMETRY_AVAILABLE = False
    trace = None  # type: ignore
    Span = None  # type: ignore
    StatusCode = None  # type: ignore
    SpanKind = None  # type: ignore

P = ParamSpec("P")
T = TypeVar("T")


# ============================================================================
# Types
# ============================================================================


@dataclass
class TracingConfig:
    """追踪配置"""

    service_name: str = "nanobot"
    """服务名称"""

    jaeger_host: str = "localhost"
    """Jaeger 主机"""

    jaeger_port: int = 6831
    """Jaeger 端口"""

    sampling_rate: float = 1.0
    """采样率 (0-1)"""

    enabled: bool = True
    """是否启用追踪"""

    console_export: bool = False
    """是否导出到控制台（调试用）"""


@dataclass
class TraceContext:
    """追踪上下文"""

    trace_id: str
    """Trace ID"""

    span_id: str
    """Span ID"""

    sampled: bool = True
    """采样标志"""


@dataclass
class TracingStats:
    """追踪统计"""

    spans_created: int = 0
    """创建的 Span 数量"""

    error_spans: int = 0
    """错误 Span 数量"""

    active_spans: int = 0
    """活跃 Span 数量"""


# ============================================================================
# TracingManager Class
# ============================================================================


class TracingManager:
    """
    追踪管理器

    管理分布式追踪功能，提供 Span 创建、上下文传播等能力

    Example:
        tracing = TracingManager(TracingConfig(service_name="my-service"))

        # 创建 Span
        async with tracing.span("operation") as span:
            span.set_attribute("key", "value")
            # 执行操作

        # 传播上下文
        headers = {}
        tracing.inject_context(headers)
    """

    def __init__(self, config: Optional[TracingConfig] = None):
        """
        初始化追踪管理器

        Args:
            config: 追踪配置，默认使用空配置
        """
        self.config = config or TracingConfig()
        self.enabled = self.config.enabled and OPENTELEMETRY_AVAILABLE
        self._tracer = None
        self._stats = TracingStats()

        if self.enabled and trace:
            self._init_tracer()

    def _init_tracer(self) -> None:
        """初始化 OpenTelemetry Tracer"""
        if not OPENTELEMETRY_AVAILABLE or not trace:
            return

        resource = Resource.create({"service.name": self.config.service_name})
        provider = TracerProvider(resource=resource)

        # 控制台导出器（调试用）
        if self.config.console_export:
            console_exporter = ConsoleSpanExporter()
            provider.add_span_processor(BatchSpanProcessor(console_exporter))

        trace.set_tracer_provider(provider)
        self._tracer = trace.get_tracer(__name__)

    def is_enabled(self) -> bool:
        """检查追踪是否启用"""
        return self.enabled

    def enable(self) -> None:
        """启用追踪"""
        self.enabled = OPENTELEMETRY_AVAILABLE

    def disable(self) -> None:
        """禁用追踪"""
        self.enabled = False

    def get_tracer(self):
        """获取 Tracer 实例"""
        if not self.enabled or not self._tracer:
            return None
        return self._tracer

    def get_active_span(self) -> Optional[Any]:
        """获取当前活跃的 Span"""
        if not self.enabled or not trace:
            return None
        return trace.get_current_span()

    @asynccontextmanager
    async def span(
        self,
        name: str,
        attributes: Optional[Dict[str, Any]] = None,
        kind: Optional[str] = None,
    ):
        """
        创建异步 Span

        Args:
            name: Span 名称
            attributes: Span 属性
            kind: Span 类型

        Yields:
            Span 实例或 Mock 对象
        """
        if not self.enabled or not self._tracer:
            yield _MockSpan()
            return

        span_kind = self._get_span_kind(kind)
        with self._tracer.start_as_current_span(name, kind=span_kind) as span:
            self._stats.spans_created += 1
            self._stats.active_spans += 1

            if attributes:
                for key, value in attributes.items():
                    span.set_attribute(key, value)

            try:
                yield span
                span.set_status(StatusCode.OK)
            except Exception as e:
                self._stats.error_spans += 1
                span.set_status(StatusCode.ERROR, str(e))
                span.record_exception(e)
                raise
            finally:
                self._stats.active_spans -= 1

    @contextmanager
    def sync_span(
        self,
        name: str,
        attributes: Optional[Dict[str, Any]] = None,
        kind: Optional[str] = None,
    ):
        """
        创建同步 Span

        Args:
            name: Span 名称
            attributes: Span 属性
            kind: Span 类型

        Yields:
            Span 实例或 Mock 对象
        """
        if not self.enabled or not self._tracer:
            yield _MockSpan()
            return

        span_kind = self._get_span_kind(kind)
        with self._tracer.start_as_current_span(name, kind=span_kind) as span:
            self._stats.spans_created += 1
            self._stats.active_spans += 1

            if attributes:
                for key, value in attributes.items():
                    span.set_attribute(key, value)

            try:
                yield span
                span.set_status(StatusCode.OK)
            except Exception as e:
                self._stats.error_spans += 1
                span.set_status(StatusCode.ERROR, str(e))
                span.record_exception(e)
                raise
            finally:
                self._stats.active_spans -= 1

    def _get_span_kind(self, kind: Optional[str]) -> Any:
        """获取 Span 类型"""
        if not SpanKind or not kind:
            return SpanKind.INTERNAL if SpanKind else None

        kind_map = {
            "internal": SpanKind.INTERNAL,
            "server": SpanKind.SERVER,
            "client": SpanKind.CLIENT,
            "producer": SpanKind.PRODUCER,
            "consumer": SpanKind.CONSUMER,
        }
        return kind_map.get(kind, SpanKind.INTERNAL)

    def inject_context(self, carrier: Dict[str, str]) -> None:
        """
        注入追踪上下文到载体

        Args:
            carrier: 载体对象 (如 HTTP headers)
        """
        if not self.enabled or not trace:
            return

        span = trace.get_current_span()
        if not span or not span.is_recording():
            return

        span_context = span.get_span_context()
        # W3C Trace Context 格式
        carrier["traceparent"] = f"00-{span_context.trace_id}-{span_context.span_id}-01"

    def extract_context(self, carrier: Dict[str, str]) -> Optional[TraceContext]:
        """
        从载体提取追踪上下文

        Args:
            carrier: 载体对象 (如 HTTP headers)

        Returns:
            追踪上下文或 None
        """
        traceparent = carrier.get("traceparent")
        if not traceparent:
            return None

        # 解析 W3C Trace Context 格式
        import re

        match = re.match(
            r"^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$",
            traceparent,
            re.IGNORECASE,
        )
        if not match:
            return None

        return TraceContext(
            trace_id=match.group(2),
            span_id=match.group(3),
            sampled=match.group(4) == "01",
        )

    def add_event(self, name: str, attributes: Optional[Dict[str, Any]] = None) -> None:
        """
        添加 Span 事件

        Args:
            name: 事件名称
            attributes: 事件属性
        """
        if not self.enabled:
            return

        span = self.get_active_span()
        if span and hasattr(span, "add_event"):
            span.add_event(name, attributes or {})

    def set_attribute(self, key: str, value: Any) -> None:
        """
        设置 Span 属性

        Args:
            key: 属性名
            value: 属性值
        """
        if not self.enabled:
            return

        span = self.get_active_span()
        if span:
            span.set_attribute(key, value)

    def record_exception(self, error: Exception) -> None:
        """
        记录异常

        Args:
            error: 异常对象
        """
        if not self.enabled:
            return

        span = self.get_active_span()
        if span and hasattr(span, "record_exception"):
            span.record_exception(error)
            span.set_status(StatusCode.ERROR, str(error))

    def get_stats(self) -> TracingStats:
        """获取追踪统计"""
        return TracingStats(
            spans_created=self._stats.spans_created,
            error_spans=self._stats.error_spans,
            active_spans=self._stats.active_spans,
        )

    def reset_stats(self) -> None:
        """重置统计"""
        self._stats = TracingStats()


# ============================================================================
# Mock Span (用于追踪禁用时)
# ============================================================================


class _MockSpan:
    """Mock Span，用于追踪禁用时"""

    def set_attribute(self, key: str, value: Any) -> None:
        pass

    def add_event(self, name: str, attributes: Optional[Dict[str, Any]] = None) -> None:
        pass

    def set_status(self, status: Any, description: Optional[str] = None) -> None:
        pass

    def record_exception(self, exception: Exception) -> None:
        pass

    def is_recording(self) -> bool:
        return False


# ============================================================================
# Global Instance
# ============================================================================

_global_tracing_manager: Optional[TracingManager] = None


def get_tracing_manager() -> TracingManager:
    """获取全局追踪管理器"""
    global _global_tracing_manager
    if _global_tracing_manager is None:
        _global_tracing_manager = TracingManager()
    return _global_tracing_manager


def set_tracing_manager(manager: TracingManager) -> None:
    """设置全局追踪管理器"""
    global _global_tracing_manager
    _global_tracing_manager = manager


# ============================================================================
# Helper Functions
# ============================================================================


@asynccontextmanager
async def span(
    name: str,
    attributes: Optional[Dict[str, Any]] = None,
    kind: Optional[str] = None,
):
    """创建追踪 Span 的便捷函数"""
    manager = get_tracing_manager()
    async with manager.span(name, attributes, kind) as s:
        yield s


def inject_trace_context(carrier: Dict[str, str]) -> None:
    """注入追踪上下文的便捷函数"""
    get_tracing_manager().inject_context(carrier)


def extract_trace_context(carrier: Dict[str, str]) -> Optional[TraceContext]:
    """提取追踪上下文的便捷函数"""
    return get_tracing_manager().extract_context(carrier)

File: automaton/src/interagent/tracing.ts
/**
 * OpenTelemetry 分布式追踪模块
 * 实现跨服务的分布式追踪能力
 *
 * @module interagent.tracing
 * @version 1.0.0
 */

import {
  trace,
  context,
  Span,
  SpanStatusCode,
  SpanOptions,
  Context,
  Tracer,
} from "@opentelemetry/api";

// ============================================================================
// Types
// ============================================================================

/**
 * 追踪配置
 */
export interface TracingConfig {
  /** 服务名称 */
  serviceName: string;
  /** Jaeger 端点 */
  jaegerEndpoint?: string;
  /** 采样率 (0-1) */
  samplingRate?: number;
  /** 是否启用追踪 */
  enabled?: boolean;
}

/**
 * Span 属性类型
 */
export type SpanAttributeValue = string | number | boolean;

/**
 * Span 创建选项
 */
export interface CreateSpanOptions {
  /** Span 属性 */
  attributes?: Record<string, SpanAttributeValue>;
  /** Span 类型 */
  kind?: "internal" | "server" | "client" | "producer" | "consumer";
  /** 父 Span */
  parent?: Span;
}

/**
 * 追踪上下文
 */
export interface TraceContext {
  /** Trace ID */
  traceId: string;
  /** Span ID */
  spanId: string;
  /** 采样标志 */
  sampled: boolean;
}

/**
 * 追踪统计
 */
export interface TracingStats {
  /** 创建的 Span 数量 */
  spansCreated: number;
  /** 错误 Span 数量 */
  errorSpans: number;
  /** 活跃 Span 数量 */
  activeSpans: number;
}

// ============================================================================
// Constants
// ============================================================================

const DEFAULT_CONFIG: Partial<TracingConfig> = {
  jaegerEndpoint: "http://localhost:14268/api/traces",
  samplingRate: 1.0,
  enabled: true,
};

const TRACER_NAME = "interagent";

// ============================================================================
// TracingManager Class
// ============================================================================

/**
 * 追踪管理器
 *
 * 管理分布式追踪功能，提供 Span 创建、上下文传播等能力
 *
 * @example
 * ```typescript
 * const tracing = new TracingManager({ serviceName: "my-service" });
 *
 * // 创建 Span
 * await tracing.withSpan("operation", async (span) => {
 *   span.setAttribute("key", "value");
 *   // 执行操作
 * });
 *
 * // 传播上下文
 * const headers = {};
 * tracing.injectContext(headers);
 * ```
 */
export class TracingManager {
  private config: TracingConfig;
  private tracer: Tracer;
  private stats: TracingStats;
  private enabled: boolean;

  constructor(config: TracingConfig) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.enabled = this.config.enabled ?? true;
    this.tracer = trace.getTracer(TRACER_NAME);
    this.stats = {
      spansCreated: 0,
      errorSpans: 0,
      activeSpans: 0,
    };
  }

  /**
   * 检查追踪是否启用
   */
  isEnabled(): boolean {
    return this.enabled;
  }

  /**
   * 启用追踪
   */
  enable(): void {
    this.enabled = true;
  }

  /**
   * 禁用追踪
   */
  disable(): void {
    this.enabled = false;
  }

  /**
   * 获取当前活跃的 Span
   */
  getActiveSpan(): Span | undefined {
    return trace.getActiveSpan();
  }

  /**
   * 创建 Span 并执行函数
   *
   * @param name - Span 名称
   * @param fn - 要执行的函数
   * @param options - Span 选项
   * @returns 函数执行结果
   */
  async withSpan<T>(
    name: string,
    fn: (span: Span) => Promise<T>,
    options: CreateSpanOptions = {}
  ): Promise<T> {
    if (!this.enabled) {
      // 追踪禁用时直接执行函数
      return fn({} as Span);
    }

    const spanOptions: SpanOptions = {
      attributes: options.attributes,
    };

    return this.tracer.startActiveSpan(name, spanOptions, async (span: Span) => {
      this.stats.spansCreated++;
      this.stats.activeSpans++;

      try {
        const result = await fn(span);
        span.setStatus({ code: SpanStatusCode.OK });
        return result;
      } catch (error) {
        this.stats.errorSpans++;
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: error instanceof Error ? error.message : "Unknown error",
        });
        span.recordException(error as Error);
        throw error;
      } finally {
        this.stats.activeSpans--;
        span.end();
      }
    });
  }

  /**
   * 创建子 Span
   *
   * @param name - Span 名称
   * @param fn - 要执行的函数
   * @param options - Span 选项
   */
  async withChildSpan<T>(
    name: string,
    fn: (span: Span) => Promise<T>,
    options: CreateSpanOptions = {}
  ): Promise<T> {
    const parentSpan = this.getActiveSpan();

    if (!parentSpan) {
      return this.withSpan(name, fn, options);
    }

    return this.withSpan(name, fn, { ...options, parent: parentSpan });
  }

  /**
   * 注入追踪上下文到载体
   *
   * @param carrier - 载体对象 (如 HTTP headers)
   */
  injectContext(carrier: Record<string, string>): void {
    if (!this.enabled) return;

    const activeSpan = this.getActiveSpan();
    if (!activeSpan) return;

    const spanContext = activeSpan.spanContext();
    // W3C Trace Context 格式
    carrier["traceparent"] = `00-${spanContext.traceId}-${spanContext.spanId}-01`;
    carrier["tracestate"] = spanContext.traceState?.serialize() || "";
  }

  /**
   * 从载体提取追踪上下文
   *
   * @param carrier - 载体对象 (如 HTTP headers)
   * @returns 追踪上下文或 null
   */
  extractContext(carrier: Record<string, string>): TraceContext | null {
    const traceparent = carrier["traceparent"];
    if (!traceparent) return null;

    // 解析 W3C Trace Context 格式
    const match = traceparent.match(/^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/i);
    if (!match) return null;

    return {
      traceId: match[2],
      spanId: match[3],
      sampled: match[4] === "01",
    };
  }

  /**
   * 添加 Span 事件
   *
   * @param name - 事件名称
   * @param attributes - 事件属性
   */
  addEvent(name: string, attributes?: Record<string, SpanAttributeValue>): void {
    const span = this.getActiveSpan();
    if (span) {
      span.addEvent(name, attributes);
    }
  }

  /**
   * 设置 Span 属性
   *
   * @param key - 属性名
   * @param value - 属性值
   */
  setAttribute(key: string, value: SpanAttributeValue): void {
    const span = this.getActiveSpan();
    if (span) {
      span.setAttribute(key, value);
    }
  }

  /**
   * 记录异常
   *
   * @param error - 异常对象
   */
  recordException(error: Error): void {
    const span = this.getActiveSpan();
    if (span) {
      span.recordException(error);
      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: error.message,
      });
    }
  }

  /**
   * 获取追踪统计
   */
  getStats(): Readonly<TracingStats> {
    return { ...this.stats };
  }

  /**
   * 重置统计
   */
  resetStats(): void {
    this.stats = {
      spansCreated: 0,
      errorSpans: 0,
      activeSpans: 0,
    };
  }
}

// ============================================================================
// Global Instance
// ============================================================================

let globalTracingManager: TracingManager | null = null;

/**
 * 获取全局追踪管理器
 */
export function getTracingManager(): TracingManager {
  if (!globalTracingManager) {
    globalTracingManager = new TracingManager({
      serviceName: "automaton",
    });
  }
  return globalTracingManager;
}

/**
 * 设置全局追踪管理器
 */
export function setTracingManager(manager: TracingManager): void {
  globalTracingManager = manager;
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * 创建追踪 Span 的便捷函数
 */
export async function withSpan<T>(
  name: string,
  fn: (span: Span) => Promise<T>,
  options?: CreateSpanOptions
): Promise<T> {
  return getTracingManager().withSpan(name, fn, options);
}

/**
 * 创建子 Span 的便捷函数
 */
export async function withChildSpan<T>(
  name: string,
  fn: (span: Span) => Promise<T>,
  options?: CreateSpanOptions
): Promise<T> {
  return getTracingManager().withChildSpan(name, fn, options);
}

/**
 * 注入追踪上下文的便捷函数
 */
export function injectTraceContext(carrier: Record<string, string>): void {
  getTracingManager().injectContext(carrier);
}

/**
 * 提取追踪上下文的便捷函数
 */
export function extractTraceContext(carrier: Record<string, string>): TraceContext | null {
  return getTracingManager().extractContext(carrier);
}

// ============================================================================
// Default Export
// ============================================================================

export default TracingManager;
